File: SHT21/Python/sht21.py
#!/usr/bin/env python3
# -*- coding: UTF-8 -*-

# Importing Lybraries
import time
import rpi_i2c
import sys, select, os


# Creating the class sht21
class sht21:

    i2c = rpi_i2c.I2C()                                       # I2C Wrapper Class (recall the class)

# Complete measurement function
    def measure(self, dev=1, scl=3, sda=2):
        self.open(dev, scl, sda)                              # Open the device
        t = self.read_temperature()                           # Get temperature
        rh = self.read_humidity()                             # Get humidity
        self.i2c.close()                                      # Close the device
        return (t, rh)

# Open and initialise the I2C port in the raspberry
    def open(self, dev=1, scl=3, sda=2):
        self.i2c.open(0x40,dev, scl, sda)                     # Open the port of the device and set the the clock and data wire
        self.i2c.write([0xFE])                                # Execute Softreset Command  (default T=14Bit RH=12)
        time.sleep(0.050)

# Temperature measurement function
    def read_temperature(self):
        self.i2c.write([0xF3])                                # Trigger T measurement (no hold master)
        time.sleep(0.066)                                     # Waiting time, typ=66ms, max=85ms @ 14Bit resolution
        data = self.i2c.read(3)
        if (self._check_crc(data, 2)):
            t = ((data[0] << 8) + data[1]) & 0xFFFC           # Set status bits to zero
            t = -46.82 + ((t * 175.72) / 65536)               # T = 46.82 + (175.72 * ST/2^16 )
            return round(t, 1)
        else:
            return None

# Humidity measurement function
    def read_humidity(self):
        self.i2c.write([0xF5])                                # Trigger RH measurement (no hold master)
        time.sleep(0.03)                                      # Waiting time, typ=22ms, max=29ms @ 12Bit resolution
        data = self.i2c.read(3)
        if (self._check_crc(data, 2)):
            rh = ((data[0] << 8) + data[1]) & 0xFFFC          # Set status bits to zero
            rh = -6 + ((125 * rh) / 65536)
            if (rh > 100): rh = 100
            return round(rh, 1)
        else:
            return None

# Close function (closes the i2c connection)
    def close(self):
        self.i2c.close()

# Checksum function (standar function in order to know when some error appear)
    def _check_crc(self, data, length):
        
        #Calculates checksum for n bytes of data and compares it with expected
        crc = 0
        for i in range(length):
            crc ^= (ord(chr(data[i])))
            for bit in range(8, 0, -1):
                if crc & 0x80:
                    crc = (crc << 1) ^ 0x131                   # CRC POLYNOMIAL
                else:
                    crc = (crc << 1)
        return True if (crc == data[length]) else False

# Taking data
if __name__ == "__main__":
    SHT21 = sht21()
    while True:
        try:
            
	    # End the process if you press any key + enter.
            if sys.stdin in select.select([sys.stdin], [], [], 0)[0]:
                line = raw_input()
                break

            #Method 1
            (t , rh) = SHT21.measure(None,3,2)
            print t, rh
            
            #Method 2
            #(temperature, humidity) = SHT21.measure(None)   # No I2C-Port/Driver --> GPIO2, GPIO3
            #print("Temperature: %s   C  Humidity: %s %%" % (temperature, humidity))
        
            time.sleep(0.1)	
	    # End the process if you press any key + enter

        except:
            print("sht21 I/O Error")

time.sleep(0.2)	

File: SHT21/C++/source/sht21.c
#include "std_c.h"
#include "i2c.h"
#include "sht21.h"
#include "raspi.h"
#include <stdio.h>



uint8	Sht21Error;
uint8 CalcSht21Crc(uint8 *data,uint8 nbrOfBytes);

//--------------------------------------------------------------------------------------------------
// Name:      SHT21_Read
// Function:  Read sensor
//            
// Parameter: -
// Return:    -
//--------------------------------------------------------------------------------------------------
uint8 SHT21_Read(int16 *temperature,uint8 *humidity)
{
	uint32	val;
	uint8	buf[4];
		
	Sht21Error = 0;
	
	//=== Softreset ==================================================
	
	I2C_Write1(0xFE);			// softreset < 15ms
	DelayMs(50);
	
	//=== Temperature =================================================

	I2C_Write1(0xF3);			// start temperature measurement
	DelayMs(260);				// Temperature@14Bit typ=66ms max=85ms
	I2C_Read(buf,3);			// read temperature data
	
	if(buf[2] == CalcSht21Crc(buf,2))  // check CRC
	{
		val = buf[0];
		val <<= 8;
		val += buf[1];
		val &= 0xFFFC;
  		  		
		*temperature = ((val * 512) / 9548);
		*temperature = ((*temperature) - 937) / 2;       
	}
	else
	{
		Sht21Error |= ERROR_SHT21_CRC_TEMP;
	}
	
	//=== Humidity ===================================================

	I2C_Write1(0xF5);			// start humidity measurement
	DelayMs(60);				// RH@12Bit typ=22ms max=20ms 
	I2C_Read(buf,3);			// read humidity data
	
  	if(buf[2] == CalcSht21Crc(buf,2))
	{	
  		val = buf[0];
  		val <<= 8;
  		val += buf[1];
  		val &= 0xFFFC;
  	  		  		
  		val = ((val * 256) / 134215) - 6;
  		*humidity = val;
	}	
	else
	{
		Sht21Error |= ERROR_SHT21_CRC_TEMP;
	}
	
	if(I2cError) Sht21Error |= ERROR_SHT21_I2C;
	
	
	
	return Sht21Error;
}

//------------------------------------------------------------------------------
// Name:  Checksum
// Function:  Return value when sensor is read it good
//            
// Parameter: -
// Return:    Value or None 
//------------------------------------------------------------------------------
uint8 CalcSht21Crc(uint8 *data,uint8 nbrOfBytes)
{
	// CRC
	//const u16t POLYNOMIAL = 0x131; //P(x)=x^8+x^5+x^4+1 = 100110001
	
	uint8 byteCtr,bit,crc;

	crc = 0;

	//calculates 8-Bit checksum with given polynomial
	for (byteCtr = 0; byteCtr < nbrOfBytes; ++byteCtr)
	{ 
		crc ^= (data[byteCtr]);
		for (bit = 8; bit > 0; --bit)
		{
			if (crc & 0x80) crc = (crc << 1) ^ 0x131;
				else 		crc = (crc << 1);
		}
	}
	return(crc);
}

//--------------------------------------------------------------------------------------------------
// Name:      PrintSht21Error
// Function:  Print error flags as readable text.
//            
// Parameter: -
// Return:    -
//--------------------------------------------------------------------------------------------------
void PrintSht21Error(void)
{	
	if(Sht21Error & ERROR_SHT21_I2C)			printf("ERROR I2C-Port\n");
	if(Sht21Error & ERROR_SHT21_CRC_TEMP)		printf("ERROR Temperature CRC\n");
	if(Sht21Error & ERROR_SHT21_CRC_HUMIDITY)	printf("ERROR Humidity CRC\n");
}

